### ISSUE TITLE
Segfault with -DXst

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5362]
Migrated from [rt.perl.org#9012](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=9012) (status was 'resolved')

Searchable as RT9012$

**Comment by p5pRT** [comment]
### From @mjdominus

#### Created by @mjdominus



This program displays a severe memory leak&#8203;:

&nbsp;    $x = "0123456789";
&nbsp;    for \(1\.\.10\) \{
&nbsp;      my @&#8203;s;
&nbsp;      $x =~ /\.\*\(\.\+?\)\(?\{push @&#8203;s\, $1\}\)\(?\!\)/;
&nbsp;    \}

&nbsp;    END \{ print STDERR "Global destruction\\n" \}

\(Probably related to the bug report I just filed\.\)

While investigating it\, I ran&#8203;:

&nbsp;        /src/bleadperl\-build/perl \-DXst /tmp/mem\.pl 

Perl died from a segmentation fault partway through the run\.  However\,

&nbsp;        /src/bleadperl\-build/perl \-DX /tmp/mem\.pl 

runs and completes normally\.


I have attached the run output below\.

plover% /src/bleadperl\-build/perl \-DXst /tmp/mem\.pl 
Pad 0x816dcb0 alloc 1 for range
Pad 0x816dcb0 sv 1
Pad 0x816dcb0 alloc 2 for range
Pad 0x816dcb0 sv 2
Pad 0x816dcb0 alloc 3 for padsv
Pad 0x8175040 alloc 1 for padsv
Pad 0x8175040 alloc 2 for push
Pad 0x816dcb0 free 1
Pad 0x816dcb0 free 2
Pad 0x8175240 alloc 1 for stringify
&nbsp;    =>  
\(/tmp/mem\.pl&#8203;:8\)    null
&nbsp;    =>  
\(/tmp/mem\.pl&#8203;:8\)    const\(PV\("Global destruction\\12"\\0\)\)
&nbsp;    =>  PV\("Global destruction\\12"\\0\)  
\(/tmp/mem\.pl&#8203;:8\)    stringify
Pad 0x8175240 sv 1
Pad 0x8175240 sv 1
Pad 0x8175240 swipe 1
Pad 0x8175240 free 1

EXECUTING\.\.\.

&nbsp;    =>  
\(/tmp/mem\.pl&#8203;:0\)    enter
&nbsp;    =>  
\(/tmp/mem\.pl&#8203;:0\)    nextstate
&nbsp;    =>  
\(/tmp/mem\.pl&#8203;:2\)    const\(PV\("0123456789"\\0\)\)
&nbsp;    =>  PV\("0123456789"\\0\)  
\(/tmp/mem\.pl&#8203;:2\)    gvsv\(main&#8203;::x\)
&nbsp;    =>  PV\("0123456789"\\0\)  UNDEF  
\(/tmp/mem\.pl&#8203;:2\)    sassign
&nbsp;    =>  PV\("0123456789"\\0\)  
\(/tmp/mem\.pl&#8203;:2\)    nextstate
&nbsp;    =>  
\(/tmp/mem\.pl&#8203;:3\)    pushmark
&nbsp;    =>  \*  
\(/tmp/mem\.pl&#8203;:3\)    const\(IV\(1\)\)
&nbsp;    =>  \*  IV\(1\)  
\(/tmp/mem\.pl&#8203;:3\)    const\(IV\(10\)\)
&nbsp;    =>  \*  IV\(1\)  IV\(10\)  
\(/tmp/mem\.pl&#8203;:3\)    gv\(main&#8203;::\_\)
&nbsp;    =>  \*  IV\(1\)  IV\(10\)  GV\(\)  
\(/tmp/mem\.pl&#8203;:3\)    enteriter
&nbsp;    =>  
\(/tmp/mem\.pl&#8203;:3\)    iter
&nbsp;    =>  SV\_YES  
\(/tmp/mem\.pl&#8203;:3\)    and
&nbsp;    =>  
\(/tmp/mem\.pl&#8203;:3\)    nextstate
&nbsp;    =>  SV\_YES  PVNV\(10\)  
\(/tmp/mem\.pl&#8203;:4\)    padav\(@&#8203;s\)
Pad 0x816dcb0 sv 3
&nbsp;    =>  SV\_YES  PVNV\(10\)  IV\(0\)  
\(/tmp/mem\.pl&#8203;:4\)    nextstate
&nbsp;    =>  SV\_YES  PVNV\(10\)  
\(/tmp/mem\.pl&#8203;:5\)    gvsv\(main&#8203;::x\)
&nbsp;    =>  SV\_YES  PVNV\(10\)  PV\("0123456789"\\0\)  
\(/tmp/mem\.pl&#8203;:5\)    match
&nbsp;    =>  SV\_YES  PVNV\(10\)  
\(/tmp/mem\.pl&#8203;:5\)    nextstate
&nbsp;    =>  SV\_YES  PVNV\(10\)  
\(\(re\_eval 1\)&#8203;:1\)    pushmark
&nbsp;    =>  SV\_YES  PVNV\(10\)  \*  
\(\(re\_eval 1\)&#8203;:1\)    padavSegmentation fault \(core dumped\)
plover%

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.6.1:

Configured by root at Sat Dec 29 11:54:59 EST 2001.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.4.2-2, archname=i586-linux
    uname='linux plover.com 2.4.2-2 #1 sun apr 8 19:37:14 edt 2001 i586 unknown '
    config_args='-des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing'
    ccversion='', gccversion='2.96 20000731 (Red Hat Linux 7.1 2.96-81)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lcrypt -lutil
    libc=/lib/libc-2.2.2.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.1:
    /usr/local/lib/perl5/5.6.1/i586-linux
    /usr/local/lib/perl5/5.6.1
    /usr/local/lib/perl5/site_perl/5.6.1/i586-linux
    /usr/local/lib/perl5/site_perl/5.6.1
    /usr/local/lib/perl5/site_perl/5.6.0/i586-linux
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.1:
    HOME=/home/mjd
    LANG=C
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/lib:/usr/lib:/usr/X11R6/lib
    LOGDIR (unset)
    PATH=/home/mjd/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/usr/games:/sbin:/usr/sbin:/usr/local/bin/X11R6:/usr/local/bin/mh:/data/mysql/bin:/usr/local/bin/pbm:/usr/local/bin/ezmlm:/home/mjd/TPI/bin:/usr/local/teTeX/bin:/usr/local/mysql/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @smpeters

> \[mjd@&#8203;plover\.com \- Wed Apr 24 06&#8203;:40&#8203;:12 2002\]&#8203;:
> 
> This is a bug report for perl from mjd@&#8203;plover\.com\,
> generated with the help of perlbug 1\.33 running under perl v5\.6\.1\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> This program displays a severe memory leak&#8203;:
> 
>     $x = "0123456789";
>     for \(1\.\.10\) \{
>       my @&#8203;s;
>       $x =~ /\.\*\(\.\+?\)\(?\{push @&#8203;s\, $1\}\)\(?\!\)/;
>     \}
> 
>     END \{ print STDERR "Global destruction\\n" \}
> 
> \(Probably related to the bug report I just filed\.\)
> 
> While investigating it\, I ran&#8203;:
> 
>         /src/bleadperl\-build/perl \-DXst /tmp/mem\.pl
> 
> Perl died from a segmentation fault partway through the run\.  However\,
> 
>         /src/bleadperl\-build/perl \-DX /tmp/mem\.pl
> 
> runs and completes normally\.
> 
> 
> I have attached the run output below\.
> 
> plover% /src/bleadperl\-build/perl \-DXst /tmp/mem\.pl
> Pad 0x816dcb0 alloc 1 for range
> Pad 0x816dcb0 sv 1
> Pad 0x816dcb0 alloc 2 for range
> Pad 0x816dcb0 sv 2
> Pad 0x816dcb0 alloc 3 for padsv
> Pad 0x8175040 alloc 1 for padsv
> Pad 0x8175040 alloc 2 for push
> Pad 0x816dcb0 free 1
> Pad 0x816dcb0 free 2
> Pad 0x8175240 alloc 1 for stringify
>     =>
> \(/tmp/mem\.pl&#8203;:8\)    null
>     =>
> \(/tmp/mem\.pl&#8203;:8\)    const\(PV\("Global destruction\\12"\\0\)\)
>     =>  PV\("Global destruction\\12"\\0\)
> \(/tmp/mem\.pl&#8203;:8\)    stringify
> Pad 0x8175240 sv 1
> Pad 0x8175240 sv 1
> Pad 0x8175240 swipe 1
> Pad 0x8175240 free 1
> 
> EXECUTING\.\.\.
> 
>     =>
> \(/tmp/mem\.pl&#8203;:0\)    enter
>     =>
> \(/tmp/mem\.pl&#8203;:0\)    nextstate
>     =>
> \(/tmp/mem\.pl&#8203;:2\)    const\(PV\("0123456789"\\0\)\)
>     =>  PV\("0123456789"\\0\)
> \(/tmp/mem\.pl&#8203;:2\)    gvsv\(main&#8203;::x\)
>     =>  PV\("0123456789"\\0\)  UNDEF
> \(/tmp/mem\.pl&#8203;:2\)    sassign
>     =>  PV\("0123456789"\\0\)
> \(/tmp/mem\.pl&#8203;:2\)    nextstate
>     =>
> \(/tmp/mem\.pl&#8203;:3\)    pushmark
>     =>  \*
> \(/tmp/mem\.pl&#8203;:3\)    const\(IV\(1\)\)
>     =>  \*  IV\(1\)
> \(/tmp/mem\.pl&#8203;:3\)    const\(IV\(10\)\)
>     =>  \*  IV\(1\)  IV\(10\)
> \(/tmp/mem\.pl&#8203;:3\)    gv\(main&#8203;::\_\)
>     =>  \*  IV\(1\)  IV\(10\)  GV\(\)
> \(/tmp/mem\.pl&#8203;:3\)    enteriter
>     =>
> \(/tmp/mem\.pl&#8203;:3\)    iter
>     =>  SV\_YES
> \(/tmp/mem\.pl&#8203;:3\)    and
>     =>
> \(/tmp/mem\.pl&#8203;:3\)    nextstate
>     =>  SV\_YES  PVNV\(10\)
> \(/tmp/mem\.pl&#8203;:4\)    padav\(@&#8203;s\)
> Pad 0x816dcb0 sv 3
>     =>  SV\_YES  PVNV\(10\)  IV\(0\)
> \(/tmp/mem\.pl&#8203;:4\)    nextstate
>     =>  SV\_YES  PVNV\(10\)
> \(/tmp/mem\.pl&#8203;:5\)    gvsv\(main&#8203;::x\)
>     =>  SV\_YES  PVNV\(10\)  PV\("0123456789"\\0\)
> \(/tmp/mem\.pl&#8203;:5\)    match
>     =>  SV\_YES  PVNV\(10\)
> \(/tmp/mem\.pl&#8203;:5\)    nextstate
>     =>  SV\_YES  PVNV\(10\)
> \(\(re\_eval 1\)&#8203;:1\)    pushmark
>     =>  SV\_YES  PVNV\(10\)  \*
> \(\(re\_eval 1\)&#8203;:1\)    padavSegmentation fault \(core dumped\)
> plover%
> 

This core dump is still around in blead@&#8203;24822\.  Below is the backtrace\.

\#0  0x0810dee7 in Perl\_debop \(my\_perl=0x82d5008\, o=0x830ab38\) at dump\.c&#8203;:1639
1639                sv = \*av\_fetch\(comppad\, o\->op\_targ\, FALSE\);
\(gdb\) bt
\#0  0x0810dee7 in Perl\_debop \(my\_perl=0x82d5008\, o=0x830ab38\) at dump\.c&#8203;:1639
\#1  0x0810d788 in Perl\_runops\_debug \(my\_perl=0x82d5008\) at dump\.c&#8203;:1597
\#2  0x081ff73d in S\_regmatch \(my\_perl=0x82d5008\, prog=0x82f5390\)
&nbsp;    at regexec\.c&#8203;:3219
\#3  0x08205c50 in S\_regmatch \(my\_perl=0x82d5008\, prog=0x82f537c\)
&nbsp;    at regexec\.c&#8203;:4078
\#4  0x0820684c in S\_regmatch \(my\_perl=0x82d5008\, prog=0x82f5374\)
&nbsp;    at regexec\.c&#8203;:4145
\#5  0x081f8dde in S\_regtry \(my\_perl=0x82d5008\, prog=0x82f5330\,
&nbsp;    startpos=0x82ecd08 "0123456789"\) at regexec\.c&#8203;:2225
\#6  0x081f61a4 in Perl\_regexec\_flags \(my\_perl=0x82d5008\, prog=0x82f5330\,
&nbsp;    stringarg=0x82ecd08 "0123456789"\, strend=0x82ecd12 ""\,
&nbsp;    strbeg=0x82ecd08 "0123456789"\, minend=0\, sv=0x82f2488\, data=0x0\,
flags=3\)
&nbsp;    at regexec\.c&#8203;:1765
\#7  0x0813d317 in Perl\_pp\_match \(my\_perl=0x82d5008\) at pp\_hot\.c&#8203;:1280
\#8  0x0810d7ed in Perl\_runops\_debug \(my\_perl=0x82d5008\) at dump\.c&#8203;:1600
\#9  0x0806aaf6 in S\_run\_body \(my\_perl=0x82d5008\, oldscope=1\) at perl\.c&#8203;:1918
\#10 0x0806a1c0 in perl\_run \(my\_perl=0x82d5008\) at perl\.c&#8203;:1847
\#11 0x0806074b in main \(argc=3\, argv=0xbfe47764\, env=0xbfe47774\)
&nbsp;    at perlmain\.c&#8203;:103

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Jul 13\, 2005 at 07&#8203;:31&#8203;:36AM \-0700\, Steve Peters via RT wrote&#8203;:
> > This program displays a severe memory leak&#8203;:
> > 
> >     $x = "0123456789";
> >     for \(1\.\.10\) \{
> >       my @&#8203;s;
> >       $x =~ /\.\*\(\.\+?\)\(?\{push @&#8203;s\, $1\}\)\(?\!\)/;
> >     \}
> > 
> >     END \{ print STDERR "Global destruction\\n" \}
> > 
> > \(Probably related to the bug report I just filed\.\)
> > 
> > While investigating it\, I ran&#8203;:
> > 
> >         /src/bleadperl\-build/perl \-DXst /tmp/mem\.pl
> > 
> > Perl died from a segmentation fault partway through the run\.  However\,
\[snip\]
> This core dump is still around in blead@&#8203;24822\.  Below is the backtrace\.

yep\, this comes under the category of "re\_eval isn't safe accessing outer
lexical vars"\. Yes another problem that will vanish when I finish my my
mythical rewrite of the re\_eval code\.

\-\- 
The optimist believes that he lives in the best of all possible worlds\.
As does the pessimist\.

**Comment by p5pRT** [comment]
### From @cpansprout

This has been fixed by the commits leading up to eb58a7e122\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Jun 14 09&#8203;:42&#8203;:32 2012\, sprout wrote&#8203;:
> This has been fixed by the commits leading up to eb58a7e122\.

Actually\, it could have been fixed in 5\.16\.  I wasn’t able to reproduce
it there\.
